File: Online_HSG/DeQui_quaylui/NQuanHau.cpp
#include <bits/stdc++.h>
using namespace std;

int n;
long long cnt = 0;
bool col[20], diag1[40], diag2[40];

void Try(int i) { // đặt hậu ở hàng i
    if (i > n) {
        cnt++;
        return;
    }
    for (int j = 1; j <= n; j++) {
        if (!col[j] && !diag1[i - j + n] && !diag2[i + j]) {
            col[j] = diag1[i - j + n] = diag2[i + j] = true;
            Try(i + 1);
            // Quay lui
            col[j] = diag1[i - j + n] = diag2[i + j] = false;
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false); cin.tie(0);
    freopen("NQuanHau.inp", "r", stdin);
    freopen("NQuanHau.out", "w", stdout);

    cin >> n;
    if (n == 2 || n == 3) {
        cout << -1;
        return 0;
    }
    Try(1);
    cout << cnt;
}

File: Online_HSG/DeQui_quaylui/NQuanHau.py
import sys

# đặt lại số mức đệ quy tối đa là 10**7
sys.setrecursionlimit(10**7)

sys.stdin = open("NQuanHau.inp", "r")
sys.stdout = open("NQuanHau.out", "w")

data = sys.stdin.read().strip().split()
if not data:
    sys.exit()

n = int(data[0])

# tương ứng với: bool col[20], diag1[40], diag2[40]
col = [False] * 21
diag1 = [False] * 41
diag2 = [False] * 41

cnt = 0

def Try(i):
    global cnt
    if i > n:
        cnt += 1
        return
    for j in range(1, n + 1):
        if (not col[j]) and (not diag1[i - j + n]) and (not diag2[i + j]):
            col[j] = True
            diag1[i - j + n] = True
            diag2[i + j] = True

            Try(i + 1)

            col[j] = False
            diag1[i - j + n] = False
            diag2[i + j] = False

if n == 2 or n == 3:
    sys.stdout.write("-1")
else:
    Try(1)
    sys.stdout.write(str(cnt))
